Drop IsLimited in favor of IsReachable

Empact opened the pull request:
These two methods have had the same meaning, but inverted, since
110b62f06992d0fb989153afff2dc3aea62a674f. Having one name for a single
concept simplifies the code.

This is a follow-up to #15051.
/cc #7553

Empact committed:
Drop IsLimited in favor of IsReachable

These two methods have had the same meaning, but inverted, since
110b62f06992d0fb989153afff2dc3aea62a674f. Having one name for a single
concept simplifies the code.

laanwj commented:
I like `IsReachable` better than `IsLimited` tbh, I think it's clearer, if we have to do this (I'm not convinced) I'd pefer to go for that.

Empact commented:
Fair enough, I agree that's clearer.

* I removed the `SetReachable` default arg because I don't think it clarifies the call sites.
* I switched from `bool[]` to `std::vector<bool>` because it enables easy initialization. `std::bitset` is another option that would work in this case, with the added bonus of bounds checking. Views on `std::bitset`?

laanwj commented:
Thanks, concept ACK.

> I switched from bool[] to std::vector<bool> because it enables easy initialization.

What is the reasoning behind this? What what difficult about the old initialization?

MarcoFalke commented on src/net.cpp:
```diff
@@ -92,7 +92,7 @@ bool fListen = true;
 CCriticalSection cs_mapLocalHost;
 std::map<CNetAddr, LocalServiceInfo> mapLocalHost GUARDED_BY(cs_mapLocalHost);
-static bool vfLimited[NET_MAX] GUARDED_BY(cs_mapLocalHost) = {};
+static std::vector<bool> nets_reachable GUARDED_BY(cs_mapLocalHost)(NET_MAX, true);
```
nit: `std::array<...> g_nets_reachable`?

promag replied:
```cpp
static auto g_nets_reachable GUARDED_BY(cs_mapLocalHost)(std::bitset<NET_MAX>{}.set());
```

promag reviewed (commented):
Concept ACK, +1 reachable.

Empact commented:
@laanwj I chose to invert the underlying variable too avoid a need to convert on the way in and out. The challenge of that is that it's surprisingly difficult to directly initialize an array of unknown size to all 1s.

As of now, using `std::bitset` with an integer initializer, plus a `static_assert` to assure the int covers all bits.

promag commented:
> As of now, using std::bitset with an integer initializer, plus a static_assert to assure the int covers all bits.

What is the problem with initializing it with `std::bitset<NET_MAX>{}.set()`?

Empact commented:
> What is the problem with initializing it with `std::bitset<NET_MAX>{}.set()`?

Switched to your approach after I wrote that - int + static_assert is nice because it's a compile-time operation, but decided that's a form of micro-optimization, not worth the cost in readability.

promag commented on src/net.cpp:
```diff
@@ -92,7 +93,7 @@ bool fListen = true;
 CCriticalSection cs_mapLocalHost;
 std::map<CNetAddr, LocalServiceInfo> mapLocalHost GUARDED_BY(cs_mapLocalHost);
-static bool vfLimited[NET_MAX] GUARDED_BY(cs_mapLocalHost) = {};
+static std::bitset<NET_MAX> g_nets_reachable GUARDED_BY(cs_mapLocalHost)(std::bitset<NET_MAX>().set());
```
nit, I still have a preference for
```cpp
static auto g_nets_reachable GUARDED_BY(cs_mapLocalHost){std::bitset<NET_MAX>().set()};
```
since it doesn't repeat the type — here the type is obvious.

Empact replied:
It's a small difference, but I included it because I think it makes the declaration more readable - after GUARDED_BY seems a bit buried if you're scanning down the page.

laanwj replied:
This is the first use of `std::bitset` in this project. It's a *small* set, the memory savings here are minimal compared to the extra code that is being introduced to handle it.
I think the original code was fine.

Empact replied:
Yeah - I think of `bitset` as just `vector<bool>` without the interface requirements of `vector`, thus able to be more constrained (e.g. fixed size). But I'll revert to the array to narrow the focus of this PR.

Empact commented:
FYI I double-checked that `NET_TEREDO`, which would be out of bounds, does not extend into these apis.

sipa commented:
utACK a7d6098edff439bc26b811b8f11852eb792e6082

Empact commented:
Dropped `std::bitset` in favor of the original array, as per https://github.com/bitcoin/bitcoin/pull/15138#discussion_r247307060

```
diff --git a/src/net.cpp b/src/net.cpp
index 95177fbae..40ec17d8e 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -37,7 +37,6 @@
 #include <miniupnpc/upnperrors.h>
 #endif
 
-#include <bitset>
 #include <unordered_map>
 
 #include <math.h>
@@ -93,7 +92,7 @@ bool fListen = true;
 bool fRelayTxes = true;
 CCriticalSection cs_mapLocalHost;
 std::map<CNetAddr, LocalServiceInfo> mapLocalHost GUARDED_BY(cs_mapLocalHost);
-static std::bitset<NET_MAX> g_nets_reachable GUARDED_BY(cs_mapLocalHost)(std::bitset<NET_MAX>().set());
+static bool vfLimited[NET_MAX] GUARDED_BY(cs_mapLocalHost) = {};
 std::string strSubVersion;
 
 limitedmap<uint256, int64_t> mapAlreadyAskedFor(MAX_INV_SZ);
@@ -258,13 +257,13 @@ void SetReachable(enum Network net, bool reachable)
     if (net == NET_UNROUTABLE || net == NET_INTERNAL)
         return;
     LOCK(cs_mapLocalHost);
-    g_nets_reachable.set(net, reachable);
+    vfLimited[net] = !reachable;
 }
 
 bool IsReachable(enum Network net)
 {
     LOCK(cs_mapLocalHost);
-    return g_nets_reachable.test(net);
+    return !vfLimited[net];
 }
 
 bool IsReachable(const CNetAddr &addr)
```

Empact commented on src/net.h:
```diff
@@ -29,6 +29,7 @@
 #include <thread>
 #include <memory>
 #include <condition_variable>
+#include <vector>
```
Originally added because I used std::vector, but left it in because vector is used in the header irrespective of other changes / iwyu. Still remove?

promag replied:
I'd say yes, since there are no ack's yet and it's unrelated to this PR.

ajtowns commented:
utACK 04a1b53661a71c240a9244e70e5f990ae767a0bf

nit: comment for `SetReachable` seems like it's a bit inverted (describes SetLimited` better)

laanwj commented on src/net.h:
```diff
@@ -520,17 +520,23 @@ enum
-bool IsLimited(const CNetAddr& addr);
+
+/**
+ * Mark a network as reachable or unreachable (no automatic connects to it)
```
I think the `SetReachable` comment is ok…?
